File: supermaster-backend/src/main/java/ar/com/leo/super_master_backend/dominio/concepto_calculo/entity/AplicaSobre.java
package ar.com.leo.super_master_backend.dominio.concepto_calculo.entity;

/**
 * Enum que representa los valores posibles para el campo aplica_sobre
 * en la tabla conceptos_calculo.
 *
 * Los valores están agrupados por etapa del cálculo de precio.
 * El nombre de cada valor indica su función y en qué etapa se aplica.
 */
public enum AplicaSobre {

    // ===== ETAPA: COSTO =====
    /**
     * Gasto que multiplica el costo base: COSTO × (1 + %/100)
     * Ejemplo: Embalaje +2%
     */
    GASTO_SOBRE_COSTO(Etapa.COSTO, NaturalezaConcepto.INFLACION),

    /**
     * Flag: usa proveedor.porcentaje para financiación.
     * Se aplica como multiplicador sobre el COSTO: COSTO × (1 + %FIN/100)
     * El concepto actúa como marcador, el valor real viene de proveedor.porcentaje.
     */
    FLAG_FINANCIACION_PROVEEDOR(Etapa.COSTO, NaturalezaConcepto.COSTO_PRODUCTO),

    // ===== ETAPA: MARGEN =====
    /**
     * Ajusta el margen sumando/restando puntos porcentuales: GAN.MIN.ML + porcentaje
     * El signo del porcentaje determina si aumenta (+) o reduce (-).
     * Ejemplo: Si GAN.MIN.ML = 60% y porcentaje = +25%, entonces ganancia = 85%
     */
    AJUSTE_MARGEN_PUNTOS(Etapa.MARGEN, NaturalezaConcepto.MARKUP),

    /**
     * Ajusta el margen proporcionalmente: GAN.MIN.ML × (1 + porcentaje/100)
     * El signo del porcentaje determina si aumenta (+) o reduce (-).
     * Ejemplo: Si GAN.MIN.ML = 60% y porcentaje = +10%, entonces ganancia = 66%
     */
    AJUSTE_MARGEN_PROPORCIONAL(Etapa.MARGEN, NaturalezaConcepto.MARKUP),

    /**
     * Flag: usa productoMargen.margenMinorista para el canal.
     * Solo actúa como marcador, el porcentaje del concepto se ignora.
     */
    FLAG_USAR_MARGEN_MINORISTA(Etapa.MARGEN, NaturalezaConcepto.MARKUP),

    /**
     * Flag: usa productoMargen.margenMayorista para el canal.
     * Solo actúa como marcador, el porcentaje del concepto se ignora.
     * Si un canal no tiene ninguno de estos conceptos, se usa margenMinorista por defecto.
     */
    FLAG_USAR_MARGEN_MAYORISTA(Etapa.MARGEN, NaturalezaConcepto.MARKUP),

    /**
     * Gasto después de calcular ganancia, pero antes de impuestos.
     * Multiplica: COSTO_CON_GANANCIA × (1 + concepto/100)
     * Ejemplo: LGELOG, LGEMKT se aplican con este tipo.
     */
    GASTO_POST_GANANCIA(Etapa.MARGEN, NaturalezaConcepto.INFLACION),

    // ===== ETAPA: IMPUESTOS =====
    /**
     * Flag: habilita la aplicación del IVA del producto para el canal.
     * Si existe este concepto, se aplica el IVA del producto.
     * Si NO existe, NO se aplica IVA (0%).
     * El porcentaje del concepto se ignora, solo actúa como habilitador.
     */
    FLAG_APLICAR_IVA(Etapa.IMPUESTOS, NaturalezaConcepto.IMPUESTO),

    /**
     * Impuesto que se suma al factor IMP: (IMP = 1 + IVA/100 + concepto/100)
     * Ejemplo: IIBB se suma directamente al factor IMP.
     */
    IMPUESTO_EN_FACTOR_IMP(Etapa.IMPUESTOS, NaturalezaConcepto.IMPUESTO),

    /**
     * Gasto después de aplicar impuestos: costoConImp × (1 + %/100)
     */
    GASTO_POST_IMPUESTOS(Etapa.IMPUESTOS, NaturalezaConcepto.INFLACION),

    // ===== ETAPA: PRECIO =====
    /**
     * Flag: incluye mla.precioEnvio para el producto.
     * El concepto actúa como marcador, el valor real viene de mlas.precio_envio.
     */
    FLAG_INCLUIR_ENVIO(Etapa.PRECIO, NaturalezaConcepto.COSTO_VENTA),

    /**
     * Concepto que se aplica como divisor sobre el PVP: PVP / (1 - %/100)
     * Ejemplo: Comisión ML -13%, embalaje que infla el PVP, etc.
     * <p>
     * <b>Naturaleza override:</b> por defecto cuenta como COSTO_VENTA (reduce ganancia).
     * Si el canal NO quiere contarlo como costo (gasto que infla el PVP pero el dueño
     * no lo asume como costo neto del canal, ej: KG_MAQ_EMB en KT GASTRO), se
     * sobreescribe la naturaleza a INFLACION en la columna {@code naturaleza}.
     * <p>
     * NOTA: Para conceptos de cuotas (con campo cuotas != NULL), se procesan de manera especial.
     */
    COMISION_SOBRE_PVP(Etapa.PRECIO, NaturalezaConcepto.COSTO_VENTA),

    /**
     * Flag: usa mla.comisionPorcentaje como comisión sobre PVP.
     * El concepto actúa como marcador, el valor real viene de mlas.comision_porcentaje.
     * Se suma a COMISION_SOBRE_PVP y se aplica como divisor: PVP / (1 - %/100)
     * <p>
     * <b>Naturaleza override:</b> por defecto cuenta como COSTO_VENTA (reduce ganancia).
     * Si el canal NO quiere contarla como costo (ej: KT GASTRO MAQUINA, donde el dueño
     * usa la comisión ML solo para inflar el PVP sin atribuirla al canal hijo), se
     * sobreescribe la naturaleza a INFLACION en la columna {@code naturaleza} del concepto.
     */
    FLAG_COMISION_ML(Etapa.PRECIO, NaturalezaConcepto.COSTO_VENTA),

    /**
     * Calcula el PVP basándose en el PVP del canal base (canalBase).
     * Si existe este concepto, se omite el cálculo normal y se usa:
     * PVP = PVP_CANAL_BASE × (1 + porcentaje/100)
     * Requiere que el canal tenga un canalBase configurado.
     * <p>
     * El factor afecta TANTO el PVP final como el ingreso del dueño
     * (los costosVenta e ingresoNeto del canal hijo se escalan con el mismo factor).
     * Ideal para canales propios donde el dueño captura el PVP completo.
     */
    CALCULO_SOBRE_CANAL_BASE(Etapa.PRECIO, NaturalezaConcepto.BASE),

    /**
     * Variante "reseller" de CALCULO_SOBRE_CANAL_BASE.
     * El factor afecta el PVP final del canal hijo, pero el ingreso del dueño
     * SE CORTA en este punto (no escala con factores no-RESELLER posteriores).
     * <p>
     * PVP_HIJO    = PVP_BASE × ∏(factores RESELLER) × ∏(factores no-RESELLER)
     * INGRESO_DUEÑO = PVP_BASE × ∏(factores RESELLER)
     * GANANCIA   = INGRESO_DUEÑO escalado - costoProducto
     * <p>
     * Caso de uso: el canal hijo es un re-vendedor que compra al precio
     * "PVP_BASE × factores RESELLER" (típicamente un descuento mayorista) y
     * agrega su propio markup (factores no-RESELLER) para vender al consumidor final.
     * Ejemplo: LIZZY GASTRO compra a mayorista×0.72 y vende a su cliente final ×1.5.
     */
    CALCULO_SOBRE_CANAL_BASE_RESELLER(Etapa.PRECIO, NaturalezaConcepto.BASE),

    // ===== ETAPA: POST_PRECIO =====
    /**
     * Costo oculto que infla el PVP como divisor adicional y cuenta como
     * costo de venta (retención que se descuenta del ingreso del dueño).
     * PVP / (1 - COSTO_OCULTO_PVP/100)
     * <p>
     * Ejemplos: ML_CO_MAQCENV / ML_CO_REP / ML_CO_MENAJE en ML, KH_CO en NUBE.
     * Originalmente llamado RECARGO_CUPON, pero el uso real es "costo oculto
     * de la plataforma" (retención adicional que no figura en la comisión).
     */
    COSTO_OCULTO_PVP(Etapa.POST_PRECIO, NaturalezaConcepto.COSTO_VENTA),

    /**
     * Descuento porcentual que reduce el PVP al final del cálculo.
     * PVP × (1 - DESCUENTO/100)
     * Ejemplo: DESCUENTO_MAQUINA
     */
    DESCUENTO_PORCENTUAL(Etapa.POST_PRECIO, NaturalezaConcepto.DESCUENTO),

    /**
     * Inflación aplicada como divisor separado sobre el PVP al final del cálculo.
     * Bucket independiente de COMISION_SOBRE_PVP.
     * PVP / (1 - INFLACION_DIVISOR_FINAL/100)
     * <p>
     * Caso de uso: precio tachado / cupón cosmético que infla el PVP sin
     * contarse como costo (el dueño se queda con la plata extra).
     * Ejemplos: ML_PRTACHADO, KH_CUPON, LG_INF.
     */
    INFLACION_DIVISOR_FINAL(Etapa.POST_PRECIO, NaturalezaConcepto.INFLACION),

    /**
     * Gasto del dueño que NO se traslada al PVP pero SÍ cuenta como costo de venta.
     * - PVP: no participa en ningún divisor ni multiplicador (el precio NO cambia).
     * - Indicadores: se suma a costosVenta (reduce ingresoNetoVendedor y ganancia).
     * Caso de uso: comisión interna del vendedor que el dueño absorbe sin
     * inflar el precio al consumidor (Excel KT GASTRO: KG_COMVEND).
     */
    GASTO_SIN_INFLAR_PVP(Etapa.POST_PRECIO, NaturalezaConcepto.COSTO_VENTA),

    /**
     * Flag: habilita la aplicación de precio inflado para el canal.
     * Si existe este concepto, se aplican los precios inflados de producto_canal_precio_inflado.
     * Si NO existe, NO se aplican precios inflados.
     * Solo actúa como habilitador, el porcentaje del concepto se ignora.
     */
    FLAG_APLICAR_PRECIO_INFLADO(Etapa.POST_PRECIO, NaturalezaConcepto.COSMETICO);

    private final Etapa etapa;
    private final NaturalezaConcepto naturalezaDefault;

    AplicaSobre(Etapa etapa, NaturalezaConcepto naturalezaDefault) {
        this.etapa = etapa;
        this.naturalezaDefault = naturalezaDefault;
    }

    /**
     * @return la etapa del cálculo de precio a la que pertenece este aplica_sobre.
     *         Asignada en la declaración del enum, así el compilador exige decisión
     *         cuando se agrega un valor nuevo.
     */
    public Etapa getEtapa() {
        return etapa;
    }

    /**
     * @return la naturaleza contable por defecto para conceptos con este aplica_sobre.
     *         Cada concepto puede sobreescribirla en su columna {@code naturaleza}
     *         (ver {@code ConceptoCalculo#getNaturalezaResolved()}).
     *         Asignada en la declaración del enum, así el compilador exige decisión
     *         cuando se agrega un valor nuevo.
     */
    public NaturalezaConcepto getNaturalezaDefault() {
        return naturalezaDefault;
    }

    /**
     * @return true si este aplica_sobre opera sobre el PVP del canal base
     *         (cualquiera de las variantes CALCULO_SOBRE_CANAL_BASE).
     */
    public boolean esCalculoSobreCanalBase() {
        return this == CALCULO_SOBRE_CANAL_BASE || this == CALCULO_SOBRE_CANAL_BASE_RESELLER;
    }

    /**
     * @return true si este aplica_sobre es un flag (marcador on/off).
     *         Los flags ignoran el porcentaje del concepto — el valor real
     *         viene del producto, MLA, proveedor, etc., o el flag solo habilita
     *         una funcionalidad sin valor numérico. El porcentaje debe quedar null.
     */
    public boolean esFlag() {
        return name().startsWith("FLAG_");
    }
}

File: supermaster-frontend/src/app/canal-formula/aplica-sobre.ts
// Metadata visual para los 20 valores de AplicaSobre.
// AplicaSobre define la MATEMÁTICA del PVP (qué etapa, qué fórmula). Esto es
// independiente de la NaturalezaConcepto, que define cómo impacta los indicadores.
// Patrón paralelo a NATURALEZAS_INFO (ver naturaleza.ts).

import type { EtapaId } from "./types";

// Badge class por etapa. Mantenido aquí en vez de derivarlo de ETAPAS_INFO
// para evitar la dependencia circular (etapas.ts re-exporta de este archivo).
// Si cambia el color de una etapa en etapas.ts, sincronizar acá también.
const BADGE_CLASS_POR_ETAPA: Record<EtapaId, string> = {
    COSTO: "bg-blue-100 text-blue-800 dark:bg-blue-500/20 dark:text-blue-200",
    MARGEN: "bg-emerald-100 text-emerald-800 dark:bg-emerald-500/20 dark:text-emerald-200",
    IMPUESTOS: "bg-amber-100 text-amber-800 dark:bg-amber-500/20 dark:text-amber-200",
    PRECIO: "bg-purple-100 text-purple-800 dark:bg-purple-500/20 dark:text-purple-200",
    POST_PRECIO: "bg-orange-100 text-orange-800 dark:bg-orange-500/20 dark:text-orange-200",
};

export type AplicaSobreId =
    // ===== ETAPA: COSTO =====
    | "GASTO_SOBRE_COSTO"
    | "FLAG_FINANCIACION_PROVEEDOR"
    // ===== ETAPA: MARGEN =====
    | "AJUSTE_MARGEN_PUNTOS"
    | "AJUSTE_MARGEN_PROPORCIONAL"
    | "FLAG_USAR_MARGEN_MINORISTA"
    | "FLAG_USAR_MARGEN_MAYORISTA"
    | "GASTO_POST_GANANCIA"
    // ===== ETAPA: IMPUESTOS =====
    | "FLAG_APLICAR_IVA"
    | "IMPUESTO_EN_FACTOR_IMP"
    | "GASTO_POST_IMPUESTOS"
    // ===== ETAPA: PRECIO =====
    | "FLAG_INCLUIR_ENVIO"
    | "COMISION_SOBRE_PVP"
    | "FLAG_COMISION_ML"
    | "CALCULO_SOBRE_CANAL_BASE"
    | "CALCULO_SOBRE_CANAL_BASE_RESELLER"
    // ===== ETAPA: POST_PRECIO =====
    | "COSTO_OCULTO_PVP"
    | "DESCUENTO_PORCENTUAL"
    | "INFLACION_DIVISOR_FINAL"
    | "GASTO_SIN_INFLAR_PVP"
    | "FLAG_APLICAR_PRECIO_INFLADO";

export interface AplicaSobreInfo {
    id: AplicaSobreId;
    /** Etiqueta legible larga (para tooltips, modales, manuales). */
    label: string;
    /** Etiqueta corta para chips/badges en tablas (≤22 chars). */
    labelCorto: string;
    /** Descripción de 1 línea para mostrar al lado del badge en la guía. */
    descripcion: string;
    /** Emoji propio del concepto (reemplaza la bandera ⚑ histórica). */
    icon: string;
    /** Etapa del cálculo a la que pertenece (determina el color de badge). */
    etapa: EtapaId;
}

export const APLICA_SOBRE_INFO: AplicaSobreInfo[] = [
    // ===== ETAPA: COSTO =====
    {
        id: "GASTO_SOBRE_COSTO",
        label: "Gasto sobre Costo",
        labelCorto: "% sobre Costo",
        descripcion: "Multiplica el costo base: COSTO × (1 + %/100).",
        icon: "📦",
        etapa: "COSTO",
    },
    {
        id: "FLAG_FINANCIACION_PROVEEDOR",
        label: "Financiación de proveedor",
        labelCorto: "Financ. Prov.",
        descripcion: "Usa el % de financiación del proveedor del producto.",
        icon: "🚩",
        etapa: "COSTO",
    },

    // ===== ETAPA: MARGEN =====
    {
        id: "AJUSTE_MARGEN_PUNTOS",
        label: "Ajuste de margen (puntos)",
        labelCorto: "Margen pts",
        descripcion: "Suma/resta puntos al margen. Ej: 60% + 25pts = 85%.",
        icon: "➕",
        etapa: "MARGEN",
    },
    {
        id: "AJUSTE_MARGEN_PROPORCIONAL",
        label: "Ajuste de margen (%)",
        labelCorto: "Margen %",
        descripcion: "Ajusta proporcionalmente: margen × (1 + %/100).",
        icon: "✖️",
        etapa: "MARGEN",
    },
    {
        id: "FLAG_USAR_MARGEN_MINORISTA",
        label: "Margen minorista",
        labelCorto: "Mg. Minorista",
        descripcion: "Usa el margen minorista del producto (por defecto si no hay flag).",
        icon: "🚩",
        etapa: "MARGEN",
    },
    {
        id: "FLAG_USAR_MARGEN_MAYORISTA",
        label: "Margen mayorista",
        labelCorto: "Mg. Mayorista",
        descripcion: "Usa el margen mayorista del producto en vez del minorista.",
        icon: "🚩",
        etapa: "MARGEN",
    },
    {
        id: "GASTO_POST_GANANCIA",
        label: "Gasto post-ganancia",
        labelCorto: "Post Ganancia",
        descripcion: "Multiplica después de la ganancia, antes de impuestos.",
        icon: "🪙",
        etapa: "MARGEN",
    },

    // ===== ETAPA: IMPUESTOS =====
    {
        id: "FLAG_APLICAR_IVA",
        label: "Aplicar IVA",
        labelCorto: "Aplicar IVA",
        descripcion: "Habilita el IVA del producto para el canal. Sin este flag, IVA = 0%.",
        icon: "🚩",
        etapa: "IMPUESTOS",
    },
    {
        id: "IMPUESTO_EN_FACTOR_IMP",
        label: "Impuesto en factor IMP",
        labelCorto: "Imp. en IMP",
        descripcion: "Se suma al factor de impuestos junto al IVA. Ej: IIBB.",
        icon: "📊",
        etapa: "IMPUESTOS",
    },
    {
        id: "GASTO_POST_IMPUESTOS",
        label: "Gasto post-impuestos",
        labelCorto: "Post Impuestos",
        descripcion: "Multiplica después de aplicar los impuestos.",
        icon: "💵",
        etapa: "IMPUESTOS",
    },

    // ===== ETAPA: PRECIO =====
    {
        id: "FLAG_INCLUIR_ENVIO",
        label: "Incluir costo de envío",
        labelCorto: "Incluir Envío",
        descripcion: "Suma el costo de envío del MLA al precio.",
        icon: "🚩",
        etapa: "PRECIO",
    },
    {
        id: "COMISION_SOBRE_PVP",
        label: "Comisión sobre PVP",
        labelCorto: "Comisión s/PVP",
        descripcion: "Divisor sobre PVP: PVP / (1 - %/100). Costo o inflación según naturaleza.",
        icon: "💸",
        etapa: "PRECIO",
    },
    {
        id: "FLAG_COMISION_ML",
        label: "Comisión MercadoLibre",
        labelCorto: "Comisión ML",
        descripcion: "Usa la comisión del MLA como divisor sobre PVP.",
        icon: "🚩",
        etapa: "PRECIO",
    },
    {
        id: "CALCULO_SOBRE_CANAL_BASE",
        label: "Cálculo sobre canal base (canal propio)",
        labelCorto: "Canal Base",
        descripcion: "PVP = PVP_base × (1 + %/100). Escala PVP e ingreso del dueño.",
        icon: "🔗",
        etapa: "PRECIO",
    },
    {
        id: "CALCULO_SOBRE_CANAL_BASE_RESELLER",
        label: "Cálculo sobre canal base (reseller)",
        labelCorto: "Canal Base (Reseller)",
        descripcion: "Variante reseller: el ingreso del dueño se 'corta' en este factor.",
        icon: "🔁",
        etapa: "PRECIO",
    },

    // ===== ETAPA: POST_PRECIO =====
    {
        id: "COSTO_OCULTO_PVP",
        label: "Costo oculto sobre PVP",
        labelCorto: "Costo Oculto s/PVP",
        descripcion: "Divisor adicional sobre PVP (retención de plataforma). Sí cuenta como costo.",
        icon: "🕳️",
        etapa: "POST_PRECIO",
    },
    {
        id: "DESCUENTO_PORCENTUAL",
        label: "Descuento porcentual",
        labelCorto: "Descuento %",
        descripcion: "Reduce el PVP: PVP × (1 - %/100). Ej: descuento máquinas.",
        icon: "🔻",
        etapa: "POST_PRECIO",
    },
    {
        id: "INFLACION_DIVISOR_FINAL",
        label: "Inflación divisor final",
        labelCorto: "Inflación Final",
        descripcion: "Bucket divisor independiente al final. Infla PVP sin contar como costo.",
        icon: "📈",
        etapa: "POST_PRECIO",
    },
    {
        id: "GASTO_SIN_INFLAR_PVP",
        label: "Gasto que no infla PVP",
        labelCorto: "Gasto sin inflar",
        descripcion: "Costo del dueño que NO se traslada al PVP pero sí cuenta como costo.",
        icon: "🫥",
        etapa: "POST_PRECIO",
    },
    {
        id: "FLAG_APLICAR_PRECIO_INFLADO",
        label: "Aplicar precio inflado",
        labelCorto: "Precio Inflado",
        descripcion: "Habilita las reglas de precio inflado para el canal.",
        icon: "🚩",
        etapa: "POST_PRECIO",
    },
];

const APLICA_SOBRE_BY_ID: Record<string, AplicaSobreInfo> = APLICA_SOBRE_INFO.reduce(
    (acc, info) => { acc[info.id] = info; return acc; },
    {} as Record<string, AplicaSobreInfo>,
);

/**
 * Devuelve la info visual del aplicaSobre. Si el valor es desconocido (caso
 * defensivo: backend agregó un valor nuevo), devuelve un fallback gris neutro
 * para no romper la UI.
 */
export function getAplicaSobreInfo(id: string | null | undefined): AplicaSobreInfo {
    if (id && APLICA_SOBRE_BY_ID[id]) return APLICA_SOBRE_BY_ID[id];
    return {
        id: (id ?? "DESCONOCIDO") as AplicaSobreId,
        label: id ?? "Desconocido",
        labelCorto: id ?? "Desconocido",
        descripcion: "Valor no reconocido — ¿el backend agregó un valor nuevo al enum AplicaSobre?",
        icon: "❓",
        etapa: "POST_PRECIO",
    };
}

/** Clase de badge derivada de la etapa del aplicaSobre. */
export function getAplicaSobreBadgeClass(id: string | null | undefined): string {
    const info = getAplicaSobreInfo(id);
    // Override: el RESELLER usa un color distintivo (fuchsia) para diferenciarlo del propio.
    if (info.id === "CALCULO_SOBRE_CANAL_BASE_RESELLER") {
        return "bg-fuchsia-100 text-fuchsia-800 dark:bg-fuchsia-500/20 dark:text-fuchsia-200";
    }
    return BADGE_CLASS_POR_ETAPA[info.etapa]
        ?? "bg-gray-100 text-gray-700 dark:bg-slate-700 dark:text-slate-300";
}

/** Backward-compat: mapa simple id → label largo (para usos legacy). */
export const APLICA_SOBRE_LABEL: Record<string, string> = APLICA_SOBRE_INFO.reduce(
    (acc, info) => { acc[info.id] = info.label; return acc; },
    {} as Record<string, string>,
);

/** Conserva la noción "es flag" (id que arranca con FLAG_) — útil para lógica. */
export const isFlag = (aplicaSobre: string): boolean => aplicaSobre.startsWith("FLAG_");
